Enhancement : Badge Component

Vrishabhsk opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
- introduce a new `component` : `Badge`

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
- https://github.com/WordPress/gutenberg/issues/61106
- Need for a `reusable component`

## Example Use Cases
### `Data-Views`

<details>
   <summary>Diff</summary>

```diff
diff --git a/packages/dataviews/src/dataviews-layouts/grid/index.tsx b/packages/dataviews/src/dataviews-layouts/grid/index.tsx
index 230ffe0dc5..cd53cd401d 100644
--- a/packages/dataviews/src/dataviews-layouts/grid/index.tsx
+++ b/packages/dataviews/src/dataviews-layouts/grid/index.tsx
@@ -13,6 +13,7 @@ import {
        Spinner,
        Flex,
        FlexItem,
+       Badge,
 } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';

@@ -111,12 +112,12 @@ function GridItem< Item >( {
                                >
                                        { badgeFields.map( ( field ) => {
                                                return (
-                                                       <FlexItem
+                                                       <Badge
                                                                key={ field.id }
                                                                className="dataviews-view-grid__field-value"
                                                        >
                                                                <field.render item={ item } />
-                                                       </FlexItem>
+                                                       </Badge>
                                                );
                                        } ) }
                                </HStack>
```

</details>

### `Post Card Panel`

<details>
    <summary>Diff</summary>

```diff
diff --git a/packages/editor/src/components/post-card-panel/index.js b/packages/editor/src/components/post-card-panel/index.js
index ed13af9b55..5bc0fef0a3 100644
--- a/packages/editor/src/components/post-card-panel/index.js
+++ b/packages/editor/src/components/post-card-panel/index.js
@@ -9,6 +9,7 @@ import {
        Icon,
        __experimentalHStack as HStack,
        __experimentalText as Text,
+       Badge,
 } from '@wordpress/components';
 import { store as coreStore } from '@wordpress/core-data';
 import { useSelect } from '@wordpress/data';
@@ -99,17 +100,12 @@ export default function PostCardPanel( {
                                        lineHeight="20px"
                                >
                                        { title ? decodeEntities( title ) : __( 'No title' ) }
-                                       { isFrontPage && (
-                                               <span className="editor-post-card-panel__title-badge">
-                                                       { __( 'Homepage' ) }
-                                               </span>
-                                       ) }
-                                       { isPostsPage && (
-                                               <span className="editor-post-card-panel__title-badge">
-                                                       { __( 'Posts Page' ) }
-                                               </span>
-                                       ) }
+                                       <Badge as="span">
+                                               { isFrontPage && __( 'Homepage' ) }
+                                               { isPostsPage && __( 'Posts Page' ) }
+                                       </Badge>
                                </Text>
+
                                <PostActions
                                        postType={ postType }
                                        postId={ postId }
```
</details>

## Screenshots <!-- if applicable -->
<img width="836" alt="Screenshot 2024-10-31 at 6 02 04 PM" src="https://github.com/user-attachments/assets/a811a969-0e97-4b92-929b-ec595ab03d73">

Vrishabhsk committed:
Create badge component

Vrishabhsk committed:
Imports and Manifest

Vrishabhsk committed:
Add test and capability for additional props using spread operator

jameskoster commented on packages/components/src/badge/styles.scss:
```diff
@@ -0,0 +1,11 @@
+	color: $gray-800;
+	padding: $grid-unit-05;
+	border-radius: $radius-small;
+	font-size: 12px;
```
We can use `$font-size-small` here.

jameskoster commented on packages/components/src/badge/styles.scss:
```diff
@@ -0,0 +1,11 @@
+	font-size: 12px;
+	font-weight: 400;
+	flex-shrink: 0;
+	line-height: $grid-unit-05 * 5;
```
`$font-line-height-small` :)

jameskoster commented on packages/components/src/badge/styles.scss:
```diff
@@ -0,0 +1,11 @@
+.components-badge {
+	background: $gray-100;
+	color: $gray-800;
+	padding: $grid-unit-05;
```
This creates a `28px` height which is `8px` larger than trunk.

An alternative approach could be to set a `min-height` and use `padding-inline`.

We might also consider increasing the `min-height` to 24px (`$grid-unit-30`) so that we can add support for icons (which are typically `24px`) later without having to change the height. 

What do you think?

Vrishabhsk replied:
SGTM 👍 

Although instead of using `padding-inline`, I just used `padding` which looked better to me

Let me know your thoughts on that.

Vrishabhsk committed:
Enhance componenet furthermore

Vrishabhsk commented:
The `tests` haven't been updated as per the latest changes. Once they are confirmed, ill update the `tests` as well

mirka commented on docs/manifest.json:
```diff
@@ -683,6 +683,12 @@
+		"slug": "badge",
+		"markdown_source": "../packages/components/src/badge/README.md",
+		"parent": "components"
+	},
```
We'll want to add this to the [ignore list](https://github.com/WordPress/gutenberg/blob/fb511a648a75bac4607f5fb9ba1f72ca5519e097/docs/tool/manifest.js#L14) while it's in private development.

mirka commented on packages/components/src/badge/README.md:
```diff
@@ -0,0 +1,59 @@
+# Badge
```
We recently added README auto-generator, so we can use that here. Add a `docs-manifest.json` file [like this](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/angle-picker-control/docs-manifest.json#L0-L1) and run `npm run docs:components`.

mirka commented on packages/components/src/badge/stories/index.story.tsx:
```diff
@@ -0,0 +1,74 @@
+
+export default meta;
+
+const Template: StoryFn< typeof Badge > = ( args ) => {
```
It would be nice if we could author new stories in [CSF 3 format](https://storybook.js.org/docs/api/csf#upgrading-from-csf-2-to-csf-3).

mirka reviewed (commented):
Thanks for taking the initiative here!

I know we're still discussing component specs in the original issue, so I'll leave the technical review for later and just leave administrative notes for now. The main thing is that this component needs to start as a private API. It needs to be a [locked export](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/private-apis.ts), and include a [`'status-private'` tag](https://github.com/WordPress/gutenberg/blob/fb511a648a75bac4607f5fb9ba1f72ca5519e097/packages/components/src/tabs/stories/index.story.tsx#L36) in the Storybook docs.

Vrishabhsk committed:
Generate README via manifest & Add in ignore list

Vrishabhsk committed:
Lock Badge

Vrishabhsk committed:
Convert Storybook from CSF 2 to CSF 3 Format

Vrishabhsk committed:
Improve the component

Vrishabhsk commented:
Hi @mirka 👋 Thanks for your input
- I have updated the `README.md` as per your instructions
- Also implemented the necessary `storybook (CSF 3)` and `private api` steps

Let me know your thoughts on it. Thanks

Vrishabhsk committed:
Merge branch 'trunk' into add/badge-component

Vrishabhsk committed:
New iteration

jameskoster commented:
Hey @Vrishabhsk what's the latest here? Has the component been updated to reflect the recent discussions in the issue? Let me know if I can help.

Vrishabhsk commented:
Hi @jameskoster 👋 
Apologies for the delay, completely slipped through.
I will handle the Icons and implement them in the component and ping back here.

Vrishabhsk committed:
Add new icons: Error and Caution

Vrishabhsk committed:
Utilize new icons

Vrishabhsk committed:
Merge branch 'trunk' into add/badge-component

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+				`components-badge--${ context }`,
+				context !== 'neutral' && 'components-badge--has-icon',
+				className
+			) }
```
Please check our [CSS conventions](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#css), it's a bit different from standard BEM 😆

mirka commented on packages/components/src/badge/types.ts:
```diff
@@ -0,0 +1,27 @@
+	 *
+	 * @default 'div'
+	 */
+	as?: ElementType;
```
I am kind of hoping we can start without an `as` prop 😬 I'm mainly afraid of people misusing this as a `button`. Would it be reasonable to keep this just a `span` for now?

cc @WordPress/gutenberg-components for second opinions.

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+function Badge( {
+	className,
+	as: Component = 'div',
+	context = 'neutral',
```
`context` has a very specific [meaning](https://react.dev/learn/passing-data-deeply-with-context) in React, so I prefer we avoid this name.

Would `variant` work? That is what we usually name things like this in this package. Or do we foresee adding general styling variants, like bordered or rounded? In that case I might suggest something like `intent` here. @WordPress/gutenberg-design

jameskoster replied:
I don't have a strong opinion about using `variant` or something else. `Intend` could work. `Tone` is another option.

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+	/**
+	 * Returns an icon based on the badge context.
+	 *
+	 * @return {JSX.Element | null} The corresponding icon for the provided context.
```
The type in the JSDoc can be omitted because the actual type annotation in the TS makes it redundant.

```suggestion
	 * @return The corresponding icon for the provided context.
```

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+	 *
+	 * @return {JSX.Element | null} The corresponding icon for the provided context.
+	 */
+	function contextBasedIcon(): JSX.Element | null {
```
```suggestion
	function contextBasedIcon() {
```

I'm actually fine with removing this explicit annotation as well (because TS can infer that).

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+				context !== 'neutral' && 'components-badge--has-icon',
+				className
+			) }
+			aria-label={ `${ context }-badge` }
```
What was the intent of this `aria-label`? It seems inappropriate here, but I want to understand the motivation in case there is a better way.

mirka commented on packages/components/src/badge/types.ts:
```diff
@@ -0,0 +1,27 @@
+	/**
+	 * Badge variant.
+	 *
+	 * @default 'generic'
```
Outdated `@default`?

mirka commented on packages/components/src/badge/stories/index.story.tsx:
```diff
@@ -0,0 +1,49 @@
+export const Default: Story = {
+	args: {
+		children: 'Code is Poetry',
+		context: 'neutral',
```
For the Default story we generally want to avoid hardcoding the default values when unnecessary. It's easier to keep the story accurate when the underlying component code changes. (`children` is fine because there's no default value)

```suggestion
```

mirka commented on packages/components/src/badge/stories/index.story.tsx:
```diff
@@ -0,0 +1,49 @@
+
+export const WithContext: Story = {
+	args: {
+		children: 'Code is Poetry',
```
```suggestion
		...Default.args,
```

This is the general convention in our package for easier composition.

mirka commented on packages/components/src/badge/stories/index.story.tsx:
```diff
@@ -0,0 +1,49 @@
+	},
+};
+
+export const WithContext: Story = {
```
I think we'll want a story for each variant, kind of similar to the Button component.

mirka commented on packages/components/src/badge/stories/index.story.tsx:
```diff
@@ -0,0 +1,49 @@
+		},
+		children: {
+			control: { type: null },
+		},
```
These can be handled by Storybook automatically, based on the type data. We want to avoid hardcoding these as much as possible, as it adds a maintenance burden.

```suggestion
	argTypes: {
		as: {
			control: { type: 'select' },
			options: [ 'div', 'span' ],
		},
```

mirka commented on packages/components/src/badge/styles.scss:
```diff
@@ -0,0 +1,38 @@
+	gap: 2px;
+
+	&--has-icon {
+		padding-left: $grid-unit-05;
```
For easier RTL support (we haven't added a lint rule for this yet, sorry).

```suggestion
		padding-inline-start: $grid-unit-05;
```

mirka commented on packages/components/src/badge/styles.scss:
```diff
@@ -0,0 +1,38 @@
+		&.components-badge--#{$type} {
+			@include badge-color-variant( $color );
+		}
+	}
```
I think it would be easier for potential overrides if we flipped this, and mixed the colors in the browser instead of in SCSS. So something like:

```diff
diff --git a/packages/components/src/badge/styles.scss b/packages/components/src/badge/styles.scss
index b0a55f9824..4cc949fdf2 100644
--- a/packages/components/src/badge/styles.scss
+++ b/packages/components/src/badge/styles.scss
@@ -1,8 +1,3 @@
-@mixin badge-color-variant( $base-color ) {
-	background-color: mix($white, $base-color, 90%);
-	color: mix($black, $base-color, 50%);
-}
-
 $badge-colors: (
 	"info": #3858e9,
 	"warning": $alert-yellow,
@@ -11,8 +6,8 @@ $badge-colors: (
 );
 
 .components-badge {
-	background: $gray-100;
-	color: $gray-800;
+	background-color: color-mix(in srgb, $white 90%, var(--base-color));
+	color: color-mix(in srgb, $black 50%, var(--base-color));
 	padding: 0 $grid-unit-10;
 	min-height: $grid-unit-30;
 	border-radius: $radius-small;
@@ -29,10 +24,15 @@ $badge-colors: (
 		padding-left: $grid-unit-05;
 	}
 
+	&:where(.is-neutral) {
+		background-color: $gray-100;
+		color: $gray-800;
+	}
+
 	// Generate color variants
 	@each $type, $color in $badge-colors {
 		&.components-badge--#{$type} {
-			@include badge-color-variant( $color );
+			--base-color: #{$color};
 		}
 	}
 }
```

mirka commented on packages/components/src/badge/styles.scss:
```diff
@@ -0,0 +1,38 @@
+@mixin badge-color-variant( $base-color ) {
+	background-color: mix($white, $base-color, 90%);
+	color: mix($black, $base-color, 50%);
```
@WordPress/gutenberg-design Have we considered what we want to happen when a Badge is placed on a dark background? (This can be a follow-up)

jameskoster replied:
What's the concern?

mirka replied:
Just thinking forward about theming. The default badge is subtle on a white background, but eye catching on a dark one. Color differences are also harder to perceive, I think, especially in isolation.

<img src="https://github.com/user-attachments/assets/3fda27b3-19e4-477a-b054-eb9e85226088" alt="Default badge" width="149">
<img src="https://github.com/user-attachments/assets/01fe7825-7662-43ed-8027-31059d3544f6" alt="Success badge" width="149">

FWIW, the default badge looks like this when rendered in the current theming system:

<img src="https://github.com/user-attachments/assets/d27e7efb-0a62-498b-b6b8-84a6b4f080c3" alt="Default badge in dark theming" width="149">

(To be clear, these issues aren't blockers for this PR to land.)

mirka commented on packages/components/src/badge/types.ts:
```diff
@@ -0,0 +1,27 @@
+	/**
+	 * Element to display inside the badge.
+	 */
+	children: ReactNode;
```
@WordPress/gutenberg-design In terms of guidelines, are there things we _don't_ want people to render inside the badge? I'm thinking about things like images, other icons, buttons, links, etc.

I'm basically wondering if we should start by only allowing plain text in the badge.

jameskoster replied:
> I'm basically wondering if we should start by only allowing plain text in the badge

Yes I think that makes sense to start.

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+function Badge( {
+	className,
+	as: Component = 'div',
+	context = 'neutral',
```
Though I do understand the benefits of using "neutral" in this context, I am leaning toward naming the default variant `"default"` 🤔 Otherwise I would have to look at the docs in order to know whether "neutral" is the default or not. What do you think?

For what it's worth, I think all our other components with variants default to either `"default"` or `undefined`.

mirka commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+	context = 'neutral',
+	children,
+	...props
+}: BadgeProps ) {
```
When passing down `...props` to the underlying element, we want to add in the underlying element props as well. We have a utility to do that (more info [here](https://wordpress.github.io/gutenberg/?path=/docs/components-contributing-guidelines--docs#typescript)):

```suggestion
}: WordPressComponentProps< BadgeProps, 'span', false > ) {
```

mirka commented on packages/components/src/index.ts:
```diff
@@ -241,6 +241,7 @@ export {
 } from './higher-order/with-focus-return';
 export { default as withNotices } from './higher-order/with-notices';
 export { default as withSpokenMessages } from './higher-order/with-spoken-messages';
+export { default as Badge } from './badge';
```
```suggestion
```

We don't want to expose this, since we locked it into a private API.

jameskoster committed:
Update icons

jameskoster commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,67 @@
+			{ context !== 'neutral' && (
+				<Icon
+					icon={ contextBasedIcon() }
+					size={ 20 }
```
There aren't many (any?) examples of icons rendering at `20px` in the software, but I don't mind it here. What do you think @jasmussen?

<img width="160" alt="Screenshot 2024-12-02 at 13 28 30" src="https://github.com/user-attachments/assets/4b2fc732-b4df-4ff1-a5e7-5ea7a373e93b">

`24px` looks a bit odd next to the smaller text:

<img width="165" alt="Screenshot 2024-12-02 at 13 29 12" src="https://github.com/user-attachments/assets/2974e9fe-295b-4678-8d3e-f71f90ab8b8d">

rogermattic replied:
+1 on the `20px` rendering! That works well for the visual consistency. 
Thank you @jameskoster for covering it and your work on these icons!


BTW In the initial designs I even had it set to `16px`.

jasmussen replied:
Hey, thanks for the ping, sorry for the belated response. 

I have semi strong opinions on this, all related to the icons being carefully designed to be displayed at 24x24 precisely, so that the _pixels render correctly_. Even if they are SVGs meant to scale, the pixels won't alias perfectly at any resolution. 

That's a long winded way of saying: if we are going to scale the icons down, we should do it according to some careful and deliberate math, so the pixels alias as well as possible. And we should do it very rarely: it should be an exception to the rule to use scaled icons, essentially, just like how we have heuristics for typography sizes. This badge can be such an exception, certainly.

So about that math: the icons are 24x24 with a 1.5px stroke width. At 20x20, the stroke width would be 1.25px. That would render blurry both at 1x screens, and at 2x screens (2.50px). If we scaled the icon to 16 (33% smaller icon), the stroke width would be exactly 1px, so it would be crisp at both 1x and 2x screens. 

How would 16x16 look in that context, can you try? If the math checks out, it should look very crisp, even if smaller. And smaller in this case would be good insofar as it looks intentional: it's smaller enough to be significant.

jameskoster replied:
Thanks for the context, we should document this, and maybe even restrict the `size` prop on `Icon` to specific values.

`16px` seems a bit small, but I'll take it over `24px`:

<img width="152" alt="Screenshot 2024-12-04 at 13 51 39" src="https://github.com/user-attachments/assets/280fbf50-cd48-428d-9d5e-244064039d0f">

jasmussen replied:
I quite like that, because it makes the icon feel like it's part of the text, almost like an in-text glyph rather than an icon. But curious what others think.

rogermattic replied:
@jameskoster @jasmussen I completely agree with the concerns raised here. The 16px icon rendered differently in Figma for me, but I agree with @jameskoster that it looks a bit off in the screenshot above. At the same time, I agree with @jasmussen that size-wise, it blends nicely with the text, which was also my goal.

The issue, though, is that the icon wasn’t designed for this size—its stroke width looks a tiny bit too thin when rescaled, and it doesn’t fill the crop area effectively.

Would it make sense to add a set of icons to the library specifically for badges at this smaller size? These could have adjusted stroke widths and proportions to better suit their context.

BTW [Google does it well](https://fonts.google.com/icons) :D

jasmussen replied:
> Would it make sense to add a set of icons to the library specifically for badges at this smaller size? These could have adjusted stroke widths and proportions to better suit their context.

This is a bit of a large undertaking, and even now we are missing a build process (#38850) so this is likely not feasible in the foreseeable future. In that vein, I'd either keep the icons at 16px size, try 18px, just use 24x24, or omit the icon entirely.

jameskoster replied:
I vote we use the current icon at 16px. The resulting strokes are exactly 1px which I agree is a little thin, but should render sharply. 

In the future it might be beneficial to redraw the icons to use stroke-width, then we can be a bit more precise/flexible. I discuss this a bit in https://github.com/WordPress/gutenberg/issues/65786.

jasmussen replied:
> The resulting strokes are exactly 1px which I agree is a little thin, but should render sharply.

Agreed. This is especially important for 1x resolution monitors, which is still [the overwhelming majority of desktop monitors](https://gs.statcounter.com/screen-resolution-stats/).

jasmussen replied:
> In the future it might be beneficial to redraw the icons to use stroke-width, then we can be a bit more precise/flexible. I discuss this a bit in https://github.com/WordPress/gutenberg/issues/65786.

This would also potentially open the door for some animation fun. Agreed this is a big undertaking.

jameskoster committed:
decrease icon size

Vrishabhsk committed:
Address feedback

mirka committed:
Merge branch 'trunk' into add/badge-component

# Conflicts:
#	packages/icons/CHANGELOG.md

mirka commented on packages/icons/src/library/caution.js:
```diff
@@ -0,0 +1,16 @@
+ */
+import { SVG, Path } from '@wordpress/primitives';
+
+const caution = (
```
@WordPress/gutenberg-design I just want to make sure we measure twice and cut once here because publishing a new icon is kind of irreversible 😄

Would it be worth renaming these so it follows the convention of `foo` + `fooFilled` like we do for the other pairs like this? For example, change the canonical name of `warning` to `cautionFilled` (maintaining back compat with an aliased export of course).

<img width="219" alt="Caution and warning icons" src="https://github.com/user-attachments/assets/9a74491f-97ff-4a3e-997b-3a1482a2295a" />

<img width="227" alt="Help icon pair" src="https://github.com/user-attachments/assets/8ed1e9f2-611d-41bb-8647-39b0c6bf3392" />

jameskoster replied:
Oh, yes, good suggestion. Just to clarify; that means any existing instances of `warning` would map to `cautionFilled`?

We should update the `cautionFilled` icon so that the details match the new `caution` icon too. Do you think it makes sense to do that in this PR, or separately?

mirka replied:
> that means any existing instances of `warning` would map to `cautionFilled`?

Correct 👍 

> Do you think it makes sense to do that in this PR, or separately?

Yes, let me pull the icon changes out of this PR so we can address it separately with proper care. I'll prepare a PR so you can push up changes to the SVG.

@Vrishabhsk Keeping this PR on hold for just a bit more. Thank you for your patience!

mirka replied:
@jameskoster Updates to the SVG can be pushed here: #67895

mirka committed:
Fix SVG formatting

mirka committed:
Fix unit test

mirka committed:
Remove unnecessary type (already included)

mirka committed:
Update readme

mirka reviewed (commented):
Thank you for the follow-ups @Vrishabhsk! I pushed some minor fixups surfaced in the failing CI checks, but otherwise I think we're pretty much ready to land this initial iteration.

Just waiting on some confirmation from the design team about icon file naming.

mirka committed:
Merge branch 'trunk' into add/badge-component

# Conflicts:
#	packages/icons/CHANGELOG.md
#	packages/icons/src/index.js

mirka committed:
Adjust icon keywords

mirka committed:
Add changelog

mirka reviewed (approved):
Ok, I think the initial iteration is ready to be merged! Thank you everyone for the collaboration 🎉

Mamaduka commented:
@Vrishabhsk, I just pushed a small follow-up for this component, which hardens conditions for icon rendering. See: #68588.
